### PR TITLE
Add graphics unit testing helpers and tests

### PR DESCRIPTION
**Pull request by ralphweng2023**
## Description
Investigates and implements strategies for unit testing SDL2-based graphics functionality in SplashKit. Previously, testing drawing procedures was awkward, had lower accuracy, and required too many lines of code.

This PR adds a reusable helper library and sample tests demonstrating:

**New Helper Functions (`graphics_test_helpers.h`):**
- `color_difference(c1, c2)` - Calculate difference between two colors (0.0-1.0 scale)
- `colors_match(c1, c2, tolerance)` - Check if colors match within tolerance
- `bitmaps_match(a, b, tolerance)` - Pixel-by-pixel bitmap comparison
- `bitmap_diff_count(a, b)` - Count differing pixels for debugging
- `verify_filled_circle(bmp, center, radius, fill_color, bg_color)` - Validate circle drawing
- `verify_filled_rectangle(bmp, rect, fill_color)` - Validate rectangle drawing

**Sample Tests (`unit_test_graphics.cpp`):**
- Tests for helper functions (color comparison, bitmap comparison)
- Drawing tests for `fill_circle`, `fill_rectangle`, `fill_ellipse`, `draw_line`
- Uses `create_bitmap` + `option_draw_to()` for offscreen rendering (no window required)
- Calls `setup_collision_mask` after drawing to bitmaps as required

This approach reduces test code from ~30 lines to ~10 lines per drawing test case, with better accuracy and maintainability.

## Type of change
- [x] New feature (helper library for testing)
- [x] New tests

## How Has This Been Tested?
Built the test project using CMake as per build instructions. All tests pass successfully.

**Automated Testing:**
```bash
cd projects/cmake
cmake --preset macOS
cmake --build build/

cd ../../bin
./skunit_tests "[graphics]"
# Result: All tests passed (28 assertions in 7 test cases)

./skunit_tests --order rand
# Result: All tests passed (1291 assertions in 86 test cases)
```

## Testing Checklist
- [x] Code compiles without errors
- [x] All 7 new graphics test cases pass (28 assertions)
- [x] All existing tests still pass (86 total test cases, 1291 assertions)
- [x] Helper functions correctly normalize color values (0-255 to 0-1 scale)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New tests have been added to prove the functionality works
- [x] Helper functions are inline for header-only usage

### CODE REVIEW

**Review by Broccoli811 — APPROVED** [review]
This PR provides a well-designed and reusable strategy for unit testing graphics functionality in SplashKit. The bitmap-based, offscreen approach significantly improves test reliability and directly addresses the issues described in the task card.

The helper functions are clearly documented and flexible, particularly the tolerance-based color comparison and bitmap matching utilities. The shape tests demonstrate clean, maintainable patterns that future contributors can easily extend when needed.

Consider briefly documenting why bitmap-based testing was chosen over window-based SDL2 rendering, to strengthen the investigation aspect of this card.

Also, the elaborate title comments are probably unneeded, if you look into other .cpp files there aren't any big section labels to signify the intention of the code blocks.

As for the new helper functions, I am unsure if adding a separate file to the unit_tests folder is appropriate or not. I'll have to ask someone else the conventions behind creating a new .h file for the purposes of a unit test.

I have tested this on my machine, and it comes out a success so no problems there.

Overall, this is a strong contribution that meaningfully improves graphics unit testing. Approved.

File: coresdk/src/test/unit_tests/unit_test_graphics.cpp [hunk 1]
```diff
@@ -1,23 +1,278 @@
...
+#include "ellipse_drawing.h"
+#include "line_drawing.h"
+#include "drawing_options.h"
+
+#include "graphics_test_helpers.h"
 
 using namespace splashkit_lib;
 
+// =============================================================================
+// Helper Function Tests
```
Comment (Broccoli811) [inline coresdk/src/test/unit_tests/unit_test_graphics.cpp:24]: This title comment is probably unneeded. To keep the styling concise, remove these.

File: coresdk/src/test/unit_tests/unit_test_graphics.cpp [hunk 2]
```diff
@@ -1,23 +1,278 @@
...
+        REQUIRE(bitmap_diff_count(bmp1, bmp3) == -1);
+        free_bitmap(bmp3);
+    }
+
+    free_bitmap(bmp1);
+    free_bitmap(bmp2);
+}
+
+// =============================================================================
+// Shape Drawing Tests
```
Comment (Broccoli811) [inline coresdk/src/test/unit_tests/unit_test_graphics.cpp:105]: This title comment is probably unneeded. To keep the styling concise, remove these.

File: coresdk/src/test/unit_tests/unit_test_graphics.cpp [hunk 3]
```diff
@@ -1,23 +1,278 @@
...
+    {
+        color pixel = get_pixel(bmp, 100, 50);
+        REQUIRE(colors_match(pixel, bg_color, 0.1));
+    }
+
+    free_bitmap(bmp);
+}
+
+// =============================================================================
+// Screen Dimension Tests
```
Comment (Broccoli811) [inline coresdk/src/test/unit_tests/unit_test_graphics.cpp:264]: This title comment is probably unneeded. To keep the styling concise, remove these.

**Review by rory-cd — CHANGES_REQUESTED** [review]
## General Information

Type of Change:
  - [x] New feature (helper library)
  - [x] New unit tests

## Code Quality

- [x] Repository: Pull Request is made to the correct repository
- [x] Readability: The code is easy to read and follow
- [x] Maintainability: This code could easily be maintained or extended in the future

## Functionality

- [ ] Correctness: The ticket called for investigation/implementation of unit testing strategies for SDL2 drawing procedures. The PR is a useful step in the right direction, however does not address the key concern outlined in the ticket. More on this below.
- [x] Impact on Existing Functionality: The impact on existing functionality has been considered and tested

## Testing

- [x] Test Coverage: Unit tests provided for new or modified code
- [x] Test Results: All tests passed

## Documentation

- [x] Documentation: Applicable inline and external documentation is updated and clear

## Pull Request Details

- [x] PR Description: The problem is clearly described
- [x] Checklist Completion: All relevant checklist items have been reviewed and completed

## Discussion

### Ticket Overview

The ticket aims to accurately test drawing procedures, rather than approximating by sampling pixels. It shows trigonometry-based sampling as an example of what to avoid, stating "*to test the drawing, it is awkward, has lower than perfect accuracy, and requires too many lines of code*".

The ticket suggests researching alternatives, and offers two potential options for testing drawing procedures:
1. Reverse engineer the dimensions of the shape using drawn pixels
2. Do a pixel-by-pixel comparison of a SplashKit-drawn bitmap to a control bitmap (drawn using [SDL code](https://splashkit.io/guides/beyond-splashkit/sdl2/getting-started-with-graphics/#installing-sdl2-and-sdl2_image))

### Test Overview
#### 🎯 "`fill_circle` draws correctly"
This test uses trigonometry-based sampling. The main loop is done in a helper library function (`verify_filled_circle`), which samples points in a circle at 80% of the radius, and then 120%. The test itself is split into sections checking:
1. Circle center point
2. A point inside the radius
3. A point outside the radius
4. Sample points in a circle at 80% of the radius, and then 120% (the helper function) 

The helper function already checks points 2 and 3, making their inclusion redundant. However, it is helpful having separate assertions for points inside and outside of the circle. Since this is the only test where the `verify_filled_circle` function is called, the function's body could be split up into sections in the test itself.

#### 🎯 "`fill_circle` with point_2d parameter works correctly"
Similar functionality to the previous test, but without checks 2 and 3. The only additional assertion from the previous test is that the center point is checked before drawing the circle. Therefore, this test appears redundant.

#### 🎯 "`fill_rectangle` draws correctly"
This test samples:
1. The center point
2. Five points in a dice-like shape (including the center point) via `verify_filled_rectangle`
3. A point in the background

Like the `fill_circle` test, the `verify_filled_rectangle` function is only called in this one test.

#### 🎯 "`fill_ellipse` draws correctly"
This test samples:
1. The center point
2. A point in the background

Therefore it is not a strong measure of an accurately drawn shape.

#### 🎯 "`draw_line` draws correctly"
As above.

### Key Issue

While there are some improvements which could be made to avoid redundant checks, or bring niche helper functions (like `verify_filled_circle`) into the `.cpp` file, the key issue is that the majority of the tests use approximate sampling techniques, rather than the more accurate suggestions offered by the ticket.

The PR includes the function `bitmaps_match`. The function is useful and well implemented, but not utilised outside of testing itself. It could be used to implement the methods suggested by the ticket.

### Changes Requested:

#### `graphics_test_helpers.h`
- Remove the encapsulating `namespace splashki_lib {}`. The helpers aren't part of the core SDK, so don't need to be part of the `splashkit_lib` namespace. This approach is used by `logging_handling.h` too.
- Move `verify_filled_circle` and `verify_filled_rectangle` to the `.cpp`, as these are specific to one test.
- `color_difference` and `colors_match` are useful functions, but they are very similar to the `pixel_is()` function written by @Broccoli811 in #148. Suggest splitting this PR, by opening a new PR just for `graphics_test_helpers.h`. The small PR could be merged quickly, allowing both this and 148 to utilise the shared helper functions.

#### `unit_test_graphics.cpp`
Use the bitmap comparison helpers in `graphics_test_helpers.h` to rewrite the tests, using one of the approaches suggested by the ticket. Either:
- Reverse engineer the shapes
- Compare with a "control bitmap" drawn using pure SDL
- A different, yet still accurate method

**Review by 222448082Ashen — APPROVED** [review]
### Detailed Feedback

*   **Architecture & Design**:
    *   The decision to use offscreen bitmaps (`create_bitmap` + `option_draw_to`) is excellent. It decouples graphics logic testing from window management and OS-specific display drivers.
    *   The helper functions (`verify_filled_circle`, `verify_filled_rectangle`) encapsulate the verification logic well, making the actual test cases in unit_test_graphics.cpp very readable and concise.

*   **Code Quality**:
    *   **Readability**: The code is clean and follows the project's `snake_case` convention. Variable names are descriptive.
    *   **Simplicity**: The sampling strategy (e.g., checking 80% radius for inside and 120% for outside in `verify_filled_circle`) is a pragmatic approach that avoids the complexity of exact pixel-perfect matching while still catching regression errors.

*   **Correctness**:
    *   The explicit call to `setup_collision_mask(bmp)` after drawing operations is correctly implemented. This is a critical detail in SplashKit to ensure the bitmap's pixel data is synchronized for `get_pixel` calls.
    *   Color comparison logic correctly normalizes 0-255 values to a 0.0-1.0 scale.

*   **Test Coverage**:
    *   The tests cover the core primitives (`fill_circle`, `fill_rectangle`, `fill_ellipse`, `draw_line`) and the helper functions themselves.
    *   Edge cases like different bitmap sizes and tolerance levels are covered.

### Checklist Verification
- [x] **Code Readability**: Helpers make tests self-documenting.
- [x] **Maintainability**: Header-only helpers are easy to include.
- [x] **Test Thoroughness**: Good coverage of primitives and helpers.
- [x] **Performance**: Sampling approach is efficient for unit tests.

Excellent work on standardizing how graphics are tested
